### PR TITLE
Highlighting multi-line comments

### LINKED ISSUE #1732: No syntax highlighting for multiple line comments
The comment on lines 2 and 3 is black instead of green

Issue has been around for a while (including on prod!) but for some reason no one quite attended to it

Incorrect syntax highlighting physically hurts my eyes 😢 - Oliver

Image

### PR DESCRIPTION
**Pull request by KesterTan**
Fixes #1732.

Before, multi-line comments were not highlighted:
Screenshot 2024-04-22 at 4 50 42 PM

With this change:
Screenshot 2024-04-22 at 4 51 06 PM

## How Has This Been Tested?
- Try out with different submissions and check that multi-line comments are now highlighted.
- Other code highlighted is not affected.
- You can still leave behind annotations etc.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] I have run rubocop and erblint for style check. If you haven't, run `overcommit --install && overcommit --sign` to use pre-commit hook for linting
- [ ] My change requires a change to the documentation, which is located at [Autolab Docs](https://docs.autolabproject.com/)
- [ ] I have updated the documentation accordingly, included in this PR

### CODE REVIEW

File: app/views/submissions/view.html.erb [hunk 1]
```diff
@@ -52,6 +52,23 @@
...
+        let commentSpans = highlightedCode.match(/<span class="hljs-comment">([^<]*)<\/span>/g);
+        let commentContents = commentSpans.map(span => span.replace(/<[^>]+>/g, ''));
+        let splitContent = commentContents.flatMap(str => str.split('\n'));
+
+        document.querySelectorAll('pre code').forEach((block, index) => {
+            if (block.textContent !== null && block.textContent !== "" && splitContent.includes(block.textContent.replace(/\n/g, ''))) {
+                block.innerHTML = `<code class='hljs-comment'>${block.textContent}</code>`;
+            }
+        });
+    });
```
[the remark this answers is not in the document]
Comment (KesterTan) [reply app/views/submissions/view.html.erb:71]: This doesnt work functionally.

File: app/assets/stylesheets/annotations.scss [hunk 2]
```diff
@@ -937,5 +937,9 @@ span > .material-icons {
   font-size: 14px !important;
   box-shadow: 0.5px 0.5px 0.5px 0.5px grey !important;
   border-color: white !important;
-} 
+}
+
+.hljs-comment span {
+  color: #008000 !important;
```
Comment (20wildmanj) [inline app/assets/stylesheets/annotations.scss:943]: Can this be a variable? see _variables.scss

File: app/views/submissions/view.html.erb [hunk 3]
```diff
@@ -52,6 +52,32 @@
       var headerPositionStr = null;
     <% end %>
     hljs.initHighlightingOnLoad();
+    document.addEventListener('DOMContentLoaded', (event) => {
```
Comment (20wildmanj) [inline app/views/submissions/view.html.erb:55]: can you add a comment explaining exactly what you are doing here? It is a little hard to understand

**Comment by 20wildmanj** [comment]
It seems that on initial load, the highlighting works, but switching to other files breaks the highlighting (the file will revert back to broken highlighting
https://github.com/autolab/Autolab/assets/25730111/45e4e530-548f-409d-9876-fec50c1781ac

File: app/assets/stylesheets/annotations.scss (hunk 2 above)
Comment (KesterTan) [reply app/assets/stylesheets/annotations.scss:943]: Okay, changed this.

File: app/views/submissions/view.html.erb (hunk 3 above)
Comment (KesterTan) [reply app/views/submissions/view.html.erb]: okay

**Comment by 20wildmanj** [comment]
Discussed offline that this PR still has some issues with highlighting, so needs a second attempt / revision.

File: app/views/submissions/view.html.erb [hunk 4]
```diff
@@ -52,6 +52,55 @@
...
+      // we wrap the content with the hljs-comment class to highlight it
+      document.querySelectorAll('pre code').forEach((block) => {
+          if (block.textContent !== null && block.textContent !== "" && splitContent.includes(block.textContent.replace(/\n/g, ''))) {
+              let escapedText = document.createTextNode(block.textContent);
+              let codeElement = document.createElement('span');
+              codeElement.className = 'hljs-comment';
+              codeElement.appendChild(escapedText);
+              block.innerHTML = '';
+              block.appendChild(codeElement);
+          }
```
[the remark this answers is not in the document]
Comment (KesterTan) [reply app/views/submissions/view.html.erb:95]: This breaks the functionality of the code.

**Comment by KesterTan** [comment]
Updated highlighting code highlights code even if we change files. 

Highlighting code also takes into account keywords such as "#include', previously these were flagged out as comments.

Screenshot 2024-05-15 at 2 42 21 PM

To check that it works as expected, check that the output is the same as that in: https://highlightjs.org/demo

File: app/assets/stylesheets/_variables.scss [hunk 5]
```diff
@@ -14,3 +14,4 @@ $autolab-white: #fff;
 $autolab-selected-gray: #f5f5f5;
 $autolab-border-gray: #f4f1f1;
 $autolab-gray-text: #676464;
+$autolab-highlight-comments: #008000;
```
Comment (20wildmanj) [inline app/assets/stylesheets/_variables.scss:17]: nit: newline

**Comment by 20wildmanj** [comment]
Screenshot 2024-05-21 at 8 57 23 AM

seems like the issue with #include getting turned into a comment also happens with #define. Functionally though the multi-line comments seem to work a lot more reliably now. Might be worth looking into updating highlight.js since it seems like the local `highlight.pack.js` was last updated in 2015, so these issues might have been fixed in later releases. Other than nit, LGTM for the most part.

**Review by 20wildmanj — APPROVED** [review]

File: app/assets/stylesheets/_variables.scss (hunk 5 above)
Comment (KesterTan) [reply app/assets/stylesheets/_variables.scss]: fixed.

File: app/views/submissions/view.html.erb [hunk 6]
```diff
@@ -52,6 +52,59 @@
...
+      });
+    }
+    window.addEventListener('DOMContentLoaded', () => {
+        highlightComments();
+    });
+    window.navigation.addEventListener("navigate", () => {
+        highlightComments();
+    });
```
[the remark this answers is not in the document]
Comment (KesterTan) [reply app/views/submissions/view.html.erb:107]: code does not work.

**Pull request merged**